Modules/Core/Common/include/itkMetaDataObject.h

```diff
@@ -130,11 +131,32 @@ class ITK_TEMPLATE_EXPORT MetaDataObject : public MetaDataObjectBase
   void
   Print(std::ostream & os) const override;
 
+  /** Returns (metaDataObject1 == metaDataObject2). */
+  friend bool
+  operator==(const Self & lhs, const Self & rhs)
+  {
+    return lhs.m_MetaDataObjectValue == rhs.m_MetaDataObjectValue;
+  }
+
+  /** Returns (metaDataObject1 != metaDataObject2). */
+  friend bool
+  operator!=(const Self & lhs, const Self & rhs)
+  {
+    return !(lhs == rhs);
+  }
+
 protected:
   MetaDataObject() = default;
   ~MetaDataObject() override = default;
 
 private:
+  /** Internal helper function used to implement operator== for MetaDataObjectBase. */
+  bool
+  Equal(const MetaDataObjectBase & metaDataObject) const override
+  {
+    return (typeid(metaDataObject) == typeid(Self)) && (*this == static_cast<const Self &>(metaDataObject));
```

Comment:
Why is this typeid comparison done over using a dynamic_cast?

Beware there are monsters here. There have been issues related to shared libraries, linkage and duplicate symbols ( typeids ) not being equal on some operating symbols (OS X), which cause these types of comparisons or dynamic_cast to fail which the object refer to the same class/lines of code but the type information is from difference complications units. I hope we have addressed they will not reoccur. However, giving this history I need to and some checks to help if this does reoccur.

Please break up this compound statement. When the typeid or dynamic_cast fails we should add an additional check to see if the objects are from the same class. The `type_info`'s `name` method may provide this check. So add if the first comparison fails:  `if ( typeid(metaDataObject).name() == typeid(Self).name() ) itkWarningMacro("...")`. This will be of great use if the linkage issue reoccurs.

Reply:
@blowekamp Thanks for your feedback. I used `typeid`, to state _explicitly_ that two meta-data-objects _only_ compare equal when their types are equal. I think the intension of `typeid(metaDataObject) == typeid(Self)` is most clear.

However, if you prefer `dynamic_cast`, that would be fine to me as well. I could certainly make it:

    const auto metaDataObjectAsSelf = dynamic_cast<const Self *>(&metaDataObject);
    return (metaDataObjectAsSelf != nullptr) && (*this == *metaDataObjectAsSelf);


Would you like that better?

Regarding those possible issues related to shared libraries, linkage and duplicate symbols, does it help that `MetaDataObject<T>` is explicitly instantiated for a specific set of types? In https://github.com/InsightSoftwareConsortium/ITK/blob/v5.2rc01/Modules/Core/Common/src/itkMetaDataObject.cxx

I would rather not put a `itkWarningMacro` in there, as I think by definitions, `operator==` should not have any side effects.

Reply:
> ```
> const auto metaDataObjectAsSelf = dynamic_cast<const Self *>(&metaDataObject);
> return (metaDataObjectAsSelf != nullptr) && (*this == *metaDataObjectAsSelf);
> ```
> 
> 
> Would you like that better?
@N-Dekker  Yes. It is more consistent with ITK, and less likely to be detected as improper use of `static_cast`.

>> I would rather not put a itkWarningMacro in there, as I think by definitions, operator== should not have any side effects.

I will improve robustness for a problematic situation that likely can still occur in some configuration of ITK. It is important that there are checks in ITK to ensure the code is robust and errouneous states and able to be tracked down.  It will not have side effect ( display warning ) unless the problem is encountered, and it will make the issue be able to tracked down.

Reply:
My idea with this PR is that `(metaDataObjectBase1 == metaDataObjectBase2)` would return `false`, and `(metaDataObjectBase1 != metaDataObjectBase2)`  would return `true`, when they refer to objects whose value type is different. The `==` and `!=` comparisons themselves would then still be valid. That's why I think a warning or error message is not necessary. OK?